src/container/cli.c

```diff
@@ -487,13 +487,131 @@ struct cont_open_args {
 	daos_handle_t		*hdlp;
 };
 
+struct pmap_refresh_cb_arg {
+	struct dc_pool		*pra_pool;
+	uint32_t		 pra_pm_ver;
+	uint32_t		 pra_retry_nr;
+};
+
+static int
+pmap_refresh_cb(tse_task_t *task, void *data)
+{
+	struct pmap_refresh_cb_arg	*cb_arg;
+	struct dc_pool			*pool;
+	uint32_t			 pm_ver;
+	uint64_t			 delay;
+	int				 rc = task->dt_result;
+
+	cb_arg = (struct pmap_refresh_cb_arg *)data;
+	pool = cb_arg->pra_pool;
+	if (rc)
+		goto out;
+
+	pm_ver = dc_pool_get_version(pool);
+	if (pm_ver < cb_arg->pra_pm_ver) {
+		if (cb_arg->pra_retry_nr > 10) {
+			/* basically this is impossible, or there is system
+			 * issue. Return EAGAIN just for code integrality.
+			 */
+			rc = -DER_AGAIN;
+			D_ERROR(DF_UUID": pmap_refresh cannot get required "
+				"version (%d:%d), try again later "DF_RC"\n",
+				DP_UUID(pool->dp_pool), pm_ver,
+				cb_arg->pra_pm_ver, DP_RC(rc));
+			goto out;
+		}
+		if (cb_arg->pra_retry_nr > 3)
+			delay = cb_arg->pra_retry_nr * 10;
+		else
+			delay = 0;
+
+		rc = tse_task_reinit_with_delay(task, delay);
+		if (rc) {
+			D_ERROR(DF_UUID": pmap_refresh version (%d:%d), resched"
+				" failed, "DF_RC"\n", DP_UUID(pool->dp_pool),
+				pm_ver, cb_arg->pra_pm_ver, DP_RC(rc));
+			goto out;
+		}
+
+		rc = dc_task_reg_comp_cb(task, pmap_refresh_cb, cb_arg,
+					 sizeof(*cb_arg));
+		if (rc) {
+			D_ERROR(DF_UUID": pmap_refresh version (%d:%d), failed "
+				"to reg_comp_cb, "DF_RC"\n",
+				DP_UUID(pool->dp_pool), pm_ver,
+				cb_arg->pra_pm_ver, DP_RC(rc));
+			goto out;
+		}
+
+		cb_arg->pra_retry_nr++;
+		D_DEBUG(DB_TRACE, DF_UUID": pmap_refresh version (%d:%d), "
+			"in %d retry\n", DP_UUID(pool->dp_pool), pm_ver,
+			cb_arg->pra_pm_ver, cb_arg->pra_retry_nr);
+		return rc;
+	}
+out:
+	if (rc)
+		D_ERROR(DF_UUID": pmap_refresh(task %p) failed, "DF_RC"\n",
+			DP_UUID(pool->dp_pool), task, DP_RC(rc));
+	dc_pool_put(pool);
+	return rc;
+}
+
+static int
+pmap_refresh(tse_task_t *task, daos_handle_t poh, uint32_t pm_ver)
+{
+	struct dc_pool			*pool;
+	struct pmap_refresh_cb_arg	 cb_arg;
+	tse_sched_t			*sched;
+	daos_pool_query_t		*pargs;
+	tse_task_t			*ptask = NULL;
+	int				 rc;
+
+	pool = dc_hdl2pool(poh);
+	if (pool == NULL)
+		return -DER_NO_HDL;
+
+	sched = (task != NULL) ? tse_task2sched(task) : NULL;
+	rc = dc_task_create(dc_pool_query, sched, NULL, &ptask);
+	if (rc != 0)
+		goto out;
+
+	pargs = dc_task_get_args(ptask);
+	pargs->poh = poh;
+	cb_arg.pra_pool = pool;
+	cb_arg.pra_pm_ver = pm_ver;
+	cb_arg.pra_retry_nr = 0;
+	rc = dc_task_reg_comp_cb(ptask, pmap_refresh_cb, &cb_arg,
+				 sizeof(cb_arg));
+	if (rc != 0)
+		goto out;
+
+	if (task) {
+		rc = dc_task_depend(task, 1, &ptask);
+		if (rc != 0)
+			goto out;
+	}
+
+	rc = dc_task_schedule(ptask, true);
+	return rc;
+
+out:
+	if (rc) {
+		dc_pool_put(pool);
```

Comment:
it looks like in the error handling case, pool might already be executed in map_refresh_cb()?

Reply:
if the task is not scheduled its comple_cb will not trigger, right? so they will not execute at same time.

Reply:
ah, right. only dc_task_schedule will. ok. thanks.